src/drivers/intel/dmic.c

```diff
@@ -1596,75 +1571,82 @@ static int dmic_probe(struct dai *dai)
 
+	dmic->nonshared = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM,
+				  sizeof(*dmic->nonshared));
+	if (!dmic->nonshared) {
+		dai_err(dai, "dmic_probe(): non-shared alloc failed");
+		return -ENOMEM;
+	}
+
+	dmic->shared = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM,
+			       sizeof(*dmic->shared));
+	if (!dmic->shared) {
+		dai_err(dai, "dmic_probe(): shared alloc failed");
+		return -ENOMEM;
+	}
+
+	/* Common shared data for all DMIC DAI instances */
+	dmic->global = platform_shared_get(&dmic_global, sizeof(dmic_global));
+
 	/* Set state, note there is no playback direction support */
-	dmic->state = COMP_STATE_READY;
+	dmic->shared->state = COMP_STATE_READY;
 
 	/* register our IRQ handler */
-	dmic->irq = interrupt_get_irq(irq, dmic_irq_name(dai));
-	if (dmic->irq < 0) {
-		ret = dmic->irq;
+	dmic->shared->irq = interrupt_get_irq(irq, dmic_irq_name(dai));
+	if (dmic->shared->irq < 0) {
+		ret = dmic->shared->irq;
 		rfree(dmic);
 		return ret;
 	}
 
-	ret = interrupt_register(dmic->irq, dmic_irq_handler, dai);
+	ret = interrupt_register(dmic->shared->irq, dmic_irq_handler, dai);
 	if (ret < 0) {
 		dai_err(dai, "dmic failed to allocate IRQ");
 		rfree(dmic);
 		return ret;
 	}
 
 	/* Initialize start sequence handler */
-	schedule_task_init_ll(&dmic->dmicwork, SOF_UUID(dmic_work_task_uuid),
+	schedule_task_init_ll(&dmic->nonshared->dmicwork, SOF_UUID(dmic_work_task_uuid),
 			      SOF_SCHEDULE_LL_TIMER,
-			      SOF_TASK_PRI_MED, dmic_work, dai, 0, 0);
+			      SOF_TASK_PRI_MED, dmic_work, dai, cpu_get_id(), 0);
 
 	/* Enable DMIC power */
 	pm_runtime_get_sync(DMIC_POW, dai->index);
 	/* Disable dynamic clock gating for dmic before touching any reg */
 	pm_runtime_get_sync(DMIC_CLK, dai->index);
-
-	interrupt_enable(dmic->irq, dai);
-
-
+	interrupt_enable(dmic->shared->irq, dai);
 	return 0;
 }
 
 static int dmic_remove(struct dai *dai)
 {
-	struct sof_ipc_dai_dmic_params **uncached_dmic_prm = cache_to_uncache(&dmic_prm[0]);
-	int uncached_dmic_active_fifos_mask = *cache_to_uncache(&dmic_active_fifos_mask);
 	struct dmic_pdata *dmic = dai_get_drvdata(dai);
-	int i;
+	uint32_t active_fifos_mask = dmic->global->active_fifos_mask;
 
 	dai_info(dai, "dmic_remove()");
 
 	/* remove scheduling */
-	schedule_task_free(&dmic->dmicwork);
+	schedule_task_free(&dmic->nonshared->dmicwork);
 
 	rfree(dai_get_drvdata(dai));
 	dai_set_drvdata(dai, NULL);
 
-	interrupt_disable(dmic->irq, dai);
-	interrupt_unregister(dmic->irq, dai);
+	interrupt_disable(dmic->shared->irq, dai);
+	interrupt_unregister(dmic->shared->irq, dai);
 
 	/* The next end tasks must be passed if another DAI FIFO still runs */
-	dai_info(dai, "dmic_remove(), dmic_active_fifos_mask = 0x%x",
-		 uncached_dmic_active_fifos_mask);
-	if (uncached_dmic_active_fifos_mask)
+	dai_info(dai, "dmic_remove(), dmic_active_fifos_mask = 0x%x", active_fifos_mask);
+
+	rfree(dmic->nonshared);
+	rfree(dmic->shared);
+	rfree(dmic);
+	if (active_fifos_mask)
 		return 0;
```

Comment:
in fact, what does it mean if during removal FIFOs are still active? We deactivate the driver but leave the hardware alone?

Reply:
The FIFO and HW resources for that FIFO only are stopped. The common HW resources must remain running. The common things (PDM bus clock, bus power, clock generator, CIC) can be shut down only when last used FIFO is being stopped.